src/net/protocol_agnostic/stream.rs

```diff
@@ -133,78 +136,101 @@ impl PaStream {
-                    })
-                    .and_then(|(utp_socket, addr)| {
-                        utp_socket
-                            .connect(&addr)
-                            .map_err(UtpRendezvousConnectError::UtpConnect)
-                            .map(PaStream::Utp)
-                    })
-            };
-
-            tcp_connect
-                .select2(utp_connect)
-                .map(|either| match either {
-                    Either::A((stream, _)) |
-                    Either::B((stream, _)) => stream,
-                })
-                .or_else(|either| match either {
-                    Either::A((tcp_error, udp_connect)) => {
-                        udp_connect
-                            .map_err(move |utp_error| {
-                                PaRendezvousConnectError::AllProtocolsFailed {
-                                    tcp: Box::new(tcp_error),
-                                    utp: Box::new(utp_error),
-                                }
-                            })
-                            .into_boxed()
-                    }
-                    Either::B((utp_error, tcp_connect)) => {
-                        tcp_connect
-                            .map_err(move |tcp_error| {
-                                PaRendezvousConnectError::AllProtocolsFailed {
-                                    tcp: Box::new(tcp_error),
-                                    utp: Box::new(utp_error),
-                                }
-                            })
-                            .into_boxed()
-                    }
+        let handle = handle.clone();
+        let tcp_connect = {
+            TcpStream::rendezvous_connect(tcp_ch_1, &handle, p2p).map(PaStream::Tcp)
+        };
+        let udp_connect = {
+            UdpSocket::rendezvous_connect(utp_ch_1, &handle, p2p)
+                .map_err(UtpRendezvousConnectError::Rendezvous)
+                .and_then(move |(udp_socket, addr)| {
+                    trace!("udp rendezvous connect succeeded.");
+                    let (utp_socket, utp_listener) = {
+                        UtpSocket::from_socket(udp_socket, &handle).map_err(
+                            UtpRendezvousConnectError::IntoUtpSocket,
+                        )?
+                    };
+                    trace!("returning utp socket.");
+                    Ok((utp_socket, utp_listener, addr))
                 })
         };
 
+
         let ret = {
             pump_channels
-                .select2(connect)
-                .map_err(|either| match either {
-                    Either::A((err, _)) |
-                    Either::B((err, _)) => err,
-                })
-                .and_then(|either| match either {
-                    Either::A(((), connect)) => connect.into_boxed(),
-                    Either::B((stream, _)) => future::ok(stream).into_boxed(),
-                })
+            .while_driving(tcp_connect)
```

Comment:
This method is already large (~140 lines), do we really want to keep expanding it?

Reply:
It could be split into smaller functions. The thing is, if we're going to have functions which are only ever meant to be called from one place inside another function then it seems clearer to me to define those functions inside the larger function. eg.

```rust
fn foo() {
    fn implementation_detail_of_foo() {
        ...
    }
    ...
    let x = implementation_detail_of_foo();
    ...
}
```

Seems clearer than

```rust
fn foo() {
    ...
    let x = implementation_detail_of_foo();
    ...
}

fn implementation_detail_of_foo() {
    ...
}
```
 Because it gives `implementation_detail_of_foo` more precise scope. This is kind of what I was aiming for when splitting the implementation of that method into separate futures with

```rust
let pump_channels = {
    ...
};

let tcp_connect = {
    ...
};

let utp_connect = {
    ...
};
```

Reply:
yeah, this kind of scoping sounds ok, but I think it hurts unit testing of those smaller functions.